### DISCUSSION TITLE
AutoFixture not working for immutable objects with tree like structure

### DISCUSSION THREAD

**Discussion by ghost** [AutoFixture/AutoFixture#1293] [Q&A]
Hi,
This small test here:
```csharp
[TestClass]
public class TreeXTests
{
    [TestMethod]
    public void ShouldCreateSmallTree()
    {
        // Arrange
        var fixture = new Fixture();
        fixture.Behaviors.Add(new OmitOnRecursionBehavior(3));
        fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList()
            .ForEach(b => fixture.Behaviors.Remove(b));

        // Act
        var em = fixture.Create<TreeX>();

        // Assert
        em.Should().NotBeNull();
    }
}
```

Works fine when I have classes defined like this:

```csharp
class TreeX
{
    public MetaX MetaX { get; set; }

    public IReadOnlyList<TreeX> Children { get; set; }

    public TreeX OtherTreePointer { get; set; }
}

class MetaX
{
    public int A { get; set; }
}
```

But when I add constructors and remove public setters:

```csharp
class TreeX
{
    public TreeX(MetaX metaX, IReadOnlyList<TreeX> children, TreeX otherTreePointer = null)
    {
        MetaX = metaX;
        Children = children;
        OtherTreePointer = otherTreePointer;
    }

    public MetaX MetaX { get; }

    public IReadOnlyList<TreeX> Children { get; }

    public TreeX OtherTreePointer { get; }
}

class MetaX
{
    public MetaX(int a)
    {
        A = a;
    }

    public int A { get; }
}
```

It fails with the following error:

> AutoFixture.ObjectCreationExceptionWithPath: AutoFixture was unable to create an instance from TreeX, most likely because it has no public constructor, is an abstract or non-public type.

Which is not really helpful.

Also I found that if I remove `OtherTreePointer` from `TreeX` then it works again:
```csharp
class TreeX
{
    public TreeX(MetaX metaX, IReadOnlyList<TreeX> children)
    {
        MetaX = metaX;
        Children = children;
    }

    public MetaX MetaX { get; }

    public IReadOnlyList<TreeX> Children { get; }
}
```

...but I really need `OtherTreePointer`. I don't want to redesign my classes and immutability is a must.

I'm not sure whether this is an issue with the library, but if not then how can I make it work?

Thanks.

AutoFixture version: `4.11.0`

**Comment by JulianBirch** [comment]
I'm guessing it's rejecting the constructor because it's recursive. Agree the error could be more helpful, but I can see the problem: how does it create the other tree node? I think it might work if you write it using overloads e.g.

```cs
public TreeX(MetaX metaX, IReadOnlyList<TreeX> children) : this(metaX, children, null)
{
}

public TreeX(MetaX metaX, IReadOnlyList<TreeX> children, TreeX otherTreePointer)
{
    MetaX = metaX;
    Children = children;
    OtherTreePointer = otherTreePointer;
}
```

**Comment by aivascu** [comment]
@bartlomiej-stys @JulianBirch this behavior is by design. AutoFixture treats any constructor that requires an instance of the same type as the declaring class, as a copy constructor. This is indeed done to avoid recursion and in the end failure. The collection in the constructor should be generated empty unless a collection was injected previously AFAIK.

Here's a link to the commit that introduced the change.
https://github.com/AutoFixture/AutoFixture/commit/74584d7d6e350aa8772fb403d419434ba275fb83

What can be done in this situations is to customize the specific tree node type to be constructed without applying the usual safety features. For this you will need to create a new constructor query that will not check for the parameters of the same type.

```cs
[Fact]
public void ShouldCreateATreeNode()
{
    var fixture = new Fixture();
    fixture.Behaviors.Add(new NullRecursionBehavior(3));
    fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList()
        .ForEach(b => fixture.Behaviors.Remove(b));
    fixture.Customizations.Add(new FilteringSpecimenBuilder(
        new MethodInvoker(new MyModestConstructorQuery()),
        new ExactTypeSpecification(typeof(TreeX))));

    var actual = fixture.Create<TreeX>();

    Assert.NotNull(actual);
    Assert.NotNull(actual.MetaX);
}

public class MyModestConstructorQuery : IMethodQuery
{
    public IEnumerable<IMethod> SelectMethods(Type type)
    {
        if (type is null) throw new ArgumentNullException(nameof(type));

        return from ci in type.GetTypeInfo().GetConstructors()
            let parameters = ci.GetParameters()
            orderby parameters.Length ascending
            select new ConstructorMethod(ci) as IMethod;
    }
}
```

Please note the example uses the `NullRecursionBehavior` instead of the `OmitRecursionBehavior` so the `MethodInvoker` wouldn't cancel the object creation.

I will add a mention about the copy constructor into the error message to avoid confusion.

Moving the issue to the discussion section.
